pypesto/hierarchical/util.py

```diff
@@ -0,0 +1,261 @@
+
+    return get_finite_quotient(
+        numerator=num,
+        denominator=den,
+        inner_parameter_type=InnerParameterType.SCALING,
+    )
+
+
+def apply_scaling(
+    scaling_value: float, sim: List[np.ndarray], mask: List[np.ndarray]
+):
+    """Apply scaling to simulations (in-place).
+
+    Parameters
+    ----------
+    scaling_value:
+        The optimal offset for the masked simulations.
+    sim:
+        All full (unmasked) simulations.
+    mask:
+        The masks that indicate the simulation subset that corresponds to the
+        `scaling_value`.
+    """
+    for i in range(len(sim)):
+        sim[i][mask[i]] = scaling_value * sim[i][mask[i]]
+
+
+def compute_optimal_offset(
+    data: List[np.ndarray],
+    sim: List[np.ndarray],
+    sigma: List[np.ndarray],
+    mask: List[np.ndarray],
+) -> float:
+    """Compute optimal offset.
+
+    Compute optimal offset for the given measurements and model outputs. See
+    https://doi.org/10.1093/bioinformatics/btz581 SI Section 3.1 for the
+    derivation.
+    """
+    # numerator, denominator
+    num, den = 0.0, 0.0
+
+    # iterate over conditions
+    for sim_i, data_i, sigma_i, mask_i in zip(sim, data, sigma, mask):
+        # extract relevant values
+        sim_x = sim_i[mask_i]
+        data_x = data_i[mask_i]
+        sigma_x = sigma_i[mask_i]
+        # update statistics
+        num += np.nansum((data_x - sim_x) / sigma_x**2)
+        den += np.nansum(1 / sigma_x**2)
+
+    return get_finite_quotient(
+        numerator=num,
+        denominator=den,
+        inner_parameter_type=InnerParameterType.OFFSET,
+    )
+
+
+def compute_optimal_offset_coupled(
+    data: List[np.ndarray],
+    sim: List[np.ndarray],
+    sigma: List[np.ndarray],
+    mask: List[np.ndarray],
+) -> float:
+    """Compute optimal offset."""
+    # numerator, denominator
+    h, recnoise, yh, y, h2 = 0.0, 0.0, 0.0, 0.0, 0.0
+
+    # iterate over conditions
+    for sim_i, data_i, sigma_i, mask_i in zip(sim, data, sigma, mask):
+        if mask_i.max(initial=False) is False:
+            continue
+        # extract relevant values
+        sim_x = sim_i[mask_i]
+        data_x = data_i[mask_i]
+        sigma_x = sigma_i[mask_i]
+        # update statistics
+        s2 = sigma_x**2
+        h += np.nansum(sim_x / s2)
+        recnoise += np.nansum(1 / s2)
+        yh += np.nansum((sim_x * data_x) / s2)
+        y += np.nansum(data_x / s2)
+        h2 += np.nansum((sim_x**2) / s2)
+
+    r1 = (yh * h) / h2
+    r2 = (h**2) / h2
+    num = y - r1
+    den = recnoise - r2
+
+    # If simulation is essentially constant, then offset and scaling
+    # have the same effect. In this case, we set the offset to a dummy value,
+    # and estimate scaling. This is because offset is computed first, by the
+    # calculator.
+    # NB: this will cause issues if e.g. a simulation with dynamics has values
+    #     on the order of the zero check here (1e-10). An attempt to avoid this
```

Comment:
```suggestion
    #     on the order of the zero check here (1e-14). An attempt to avoid this
```

Reply:
Removed the specific value, as it's in the code.